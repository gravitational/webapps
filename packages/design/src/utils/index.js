/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import ReactDOM from 'react-dom';
import getOffset from 'dom-helpers/query/offset';
import getPosition from 'dom-helpers/query/position';
import getScrollTop from 'dom-helpers/query/scrollTop';

export function ownerDocument(node) {
  return (node && node.ownerDocument) || document;
}

export function ownerWindow(node, fallback = window) {
  const doc = ownerDocument(node);
  return doc.defaultView || doc.parentView || fallback;
}

export function getContainer(container, defaultContainer) {
  container = typeof container === 'function' ? container() : container;
  return ReactDOM.findDOMNode(container) || defaultContainer;
}

function getContainerDimensions(containerNode) {
  let width, height, scroll;

  if (containerNode.tagName === 'BODY') {
    width = window.innerWidth;
    height = window.innerHeight;

    scroll =
      getScrollTop(ownerDocument(containerNode).documentElement) ||
      getScrollTop(containerNode);
  } else {
    ({ width, height } = getOffset(containerNode));
    scroll = getScrollTop(containerNode);
  }

  return { width, height, scroll };
}

function getTopDelta(top, overlayHeight, container, padding) {
  const containerDimensions = getContainerDimensions(container);
  const containerScroll = containerDimensions.scroll;
  const containerHeight = containerDimensions.height;

  const topEdgeOffset = top - padding - containerScroll;
  const bottomEdgeOffset = top + padding - containerScroll + overlayHeight;

  if (topEdgeOffset < 0) {
    return -topEdgeOffset;
  } else if (bottomEdgeOffset > containerHeight) {
    return containerHeight - bottomEdgeOffset;
  } else {
    return 0;
  }
}

function getLeftDelta(left, overlayWidth, container, padding) {
  const containerDimensions = getContainerDimensions(container);
  const containerWidth = containerDimensions.width;

  const leftEdgeOffset = left - padding;
  const rightEdgeOffset = left + padding + overlayWidth;

  if (leftEdgeOffset < 0) {
    return -leftEdgeOffset;
  } else if (rightEdgeOffset > containerWidth) {
    return containerWidth - rightEdgeOffset;
  }

  return 0;
}

export function calculatePosition(
  placement,
  overlayNode,
  target,
  container,
  padding
) {
  const childOffset =
    container.tagName === 'BODY'
      ? getOffset(target)
      : getPosition(target, container);

  const { height: overlayHeight, width: overlayWidth } = getOffset(overlayNode);

  let positionLeft, positionTop, arrowOffsetLeft, arrowOffsetTop;

  if (placement === 'left' || placement === 'right') {
    positionTop = childOffset.top + (childOffset.height - overlayHeight) / 2;

    if (placement === 'left') {
      positionLeft = childOffset.left - overlayWidth;
    } else {
      positionLeft = childOffset.left + childOffset.width;
    }

    const topDelta = getTopDelta(
      positionTop,
      overlayHeight,
      container,
      padding
    );

    positionTop += topDelta;
    arrowOffsetTop = 50 * (1 - (2 * topDelta) / overlayHeight) + '%';
    arrowOffsetLeft = void 0;
  } else if (placement === 'top' || placement === 'bottom') {
    positionLeft = childOffset.left + (childOffset.width - overlayWidth) / 2;

    if (placement === 'top') {
      positionTop = childOffset.top - overlayHeight;
    } else {
      positionTop = childOffset.top + childOffset.height;
    }

    const leftDelta = getLeftDelta(
      positionLeft,
      overlayWidth,
      container,
      padding
    );

    positionLeft += leftDelta;
    arrowOffsetLeft = 50 * (1 - (2 * leftDelta) / overlayWidth) + '%';
    arrowOffsetTop = void 0;
  } else {
    throw new Error(
      `calcOverlayPosition(): No such placement of "${placement}" found.`
    );
  }

  return { positionLeft, positionTop, arrowOffsetLeft, arrowOffsetTop };
}
