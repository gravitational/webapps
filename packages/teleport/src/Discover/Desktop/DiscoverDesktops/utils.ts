import { Path } from 'teleport/Discover/Desktop/DiscoverDesktops/lines/Path';
import { MonotoneX } from 'teleport/Discover/Desktop/DiscoverDesktops/lines/MonotoneX';

const STROKE_WIDTH = 4;

export interface Line {
  width: number;
  height: number;
  path: string;
}

export function createLine(
  desktopServiceElement: HTMLDivElement,
  desktopElement: HTMLDivElement,
  containerElement: HTMLDivElement
): Line {
  if (!desktopElement || !desktopServiceElement || !containerElement) {
    return null;
  }

  const desktopServiceRect = desktopServiceElement.getBoundingClientRect();
  const desktopRect = desktopElement.getBoundingClientRect();
  const containerRect = containerElement.getBoundingClientRect();

  const distance = desktopRect.left - desktopServiceRect.right;

  const path = new Path();
  const line = new MonotoneX(path);

  line.lineStart();

  const desktopLinePosition =
    desktopRect.top - containerRect.top + desktopRect.height / 2 - 1;
  const desktopServiceLinePosition =
    desktopServiceRect.top - containerRect.top + desktopServiceRect.height / 2;

  line.point(0, desktopServiceLinePosition - STROKE_WIDTH * 2);
  line.point(40, desktopServiceLinePosition - STROKE_WIDTH * 2);
  line.point(distance - 10, desktopLinePosition + STROKE_WIDTH / 2);
  line.point(distance, desktopLinePosition + STROKE_WIDTH / 2);

  line.lineEnd();

  return {
    width: distance,
    height: containerRect.height,
    path: path.path,
  };
}
