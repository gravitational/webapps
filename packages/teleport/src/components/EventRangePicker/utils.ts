import moment from 'moment';

export function getRangeOptions(): EventRange[] {
  return [
    {
      name: 'Today',
      from: moment(new Date())
        .startOf('day')
        .toDate(),
      to: moment(new Date())
        .endOf('day')
        .toDate(),
    },
    {
      name: '7 days',
      from: moment()
        .subtract(6, 'day')
        .startOf('day')
        .toDate(),
      to: moment(new Date())
        .endOf('day')
        .toDate(),
    },
    {
      name: 'Custom Range...',
      isCustom: true,
      from: new Date(),
      to: new Date(),
    },
  ];
}

export type EventRange = {
  from: Date;
  to: Date;
  isCustom?: boolean;
  name?: string;
};
