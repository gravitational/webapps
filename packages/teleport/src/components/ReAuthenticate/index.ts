import ReAuthenticate from './ReAuthenticate';

export default ReAuthenticate;
