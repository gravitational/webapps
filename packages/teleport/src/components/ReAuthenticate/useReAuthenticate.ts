/*
Copyright 2021-2022 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import cfg from 'teleport/config';
import auth from 'teleport/services/auth';
import useAttempt from 'shared/hooks/useAttemptNext';

export default function useReAuthenticate({ onAuthenticated, onClose }: Props) {
  const { attempt, setAttempt, handleError } = useAttempt('');

  function submitWithTotp(secondFactorToken: string) {
    setAttempt({ status: 'processing' });
    auth
      .createPrivilegeTokenWithTotp(secondFactorToken)
      .then(onAuthenticated)
      .catch(handleError);
  }

  function submitWithWebauthn() {
    setAttempt({ status: 'processing' });
    auth
      .createPrivilegeTokenWithWebauthn()
      .then(onAuthenticated)
      .catch(handleError);
  }

  function clearAttempt() {
    setAttempt({ status: '' });
  }

  return {
    attempt,
    clearAttempt,
    submitWithTotp,
    submitWithWebauthn,
    auth2faType: cfg.getAuth2faType(),
    preferredMfaType: cfg.getPreferredMfaType(),
    onClose,
  };
}

export type Props = {
  onAuthenticated: React.Dispatch<React.SetStateAction<string>>;
  onClose: () => void;
};

export type State = ReturnType<typeof useReAuthenticate>;
