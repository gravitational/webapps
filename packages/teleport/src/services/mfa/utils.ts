/**
 * Copyright 2021 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Auth2faType, PreferredMfaType } from 'shared/services';

export type MfaOption = {
  value: Auth2faType;
  label: string;
};

export function getMfaOptions(
  mfa: Auth2faType,
  preferredMfa: PreferredMfaType,
  requireMfa = false
) {
  const mfaOptions: MfaOption[] = [];

  if (mfa === 'off' || !mfa) {
    return mfaOptions;
  }

  const mfaEnabled = mfa === 'on' || mfa === 'optional';
  const preferMfaWebauthn = mfaEnabled && preferredMfa === 'webauthn';
  const preferMfaU2f = mfaEnabled && preferredMfa === 'u2f';

  if (mfa === 'webauthn' || preferMfaWebauthn) {
    mfaOptions.push({ value: 'webauthn', label: 'Hardware Key' });
  }

  if (mfa === 'u2f' || preferMfaU2f) {
    mfaOptions.push({ value: 'u2f', label: 'Hardware Key' });
  }

  if (mfa === 'otp' || mfaEnabled) {
    mfaOptions.push({ value: 'otp', label: 'Authenticator App' });
  }

  if (!requireMfa && mfa === 'optional') {
    mfaOptions.push({ value: 'optional', label: 'None' });
  }

  return mfaOptions;
}
