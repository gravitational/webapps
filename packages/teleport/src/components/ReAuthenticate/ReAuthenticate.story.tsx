import React from 'react';
import { State } from './useReAuthenticate';
import { ReAuthenticate } from './ReAuthenticate';

export default {
  title: 'Teleport/ReAuthenticate',
};

export const Loaded = () => <ReAuthenticate {...props} />;

export const Processing = () => (
  <ReAuthenticate {...props} attempt={{ status: 'processing' }} />
);

export const Failed = () => (
  <ReAuthenticate
    {...props}
    attempt={{ status: 'failed', statusText: 'an error has occurred' }}
  />
);

const props: State = {
  attempt: { status: '' },
  clearAttempt: () => null,
  submitWithTotp: () => null,
  submitWithU2f: () => null,
  submitWithWebauthn: () => null,
  preferredMfaType: 'webauthn',
  close: () => null,
  auth2faType: 'on',
};
