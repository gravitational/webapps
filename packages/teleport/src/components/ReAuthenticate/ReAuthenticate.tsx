/*
Copyright 2021-2022 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { useState } from 'react';
import { Flex, Box, Text, ButtonPrimary, ButtonSecondary } from 'design';
import Dialog, {
  DialogHeader,
  DialogTitle,
  DialogContent,
  DialogFooter,
} from 'design/Dialog';
import { Danger } from 'design/Alert';
import Validation from 'shared/components/Validation';
import { requiredToken } from 'shared/components/Validation/rules';
import FieldInput from 'shared/components/FieldInput';
import FieldSelect from 'shared/components/FieldSelect';
import createMfaOptions, { MfaOption } from 'shared/utils/createMfaOptions';
import useReAuthenticate, { State, Props } from './useReAuthenticate';

export default function Container(props: Props) {
  const state = useReAuthenticate(props);
  return <ReAuthenticate {...state} />;
}

export function ReAuthenticate({
  attempt,
  clearAttempt,
  submitWithTotp,
  submitWithWebauthn,
  onClose,
  auth2faType,
  preferredMfaType,
}: State) {
  const [otpToken, setOtpToken] = useState('');
  const mfaOptions = createMfaOptions({
    auth2faType: auth2faType,
    preferredType: preferredMfaType,
    required: true,
  });
  const [mfaOption, setMfaOption] = useState<MfaOption>(mfaOptions[0]);

  function onSubmit(e: React.MouseEvent<HTMLButtonElement>) {
    e.preventDefault();

    if (mfaOption?.value === 'webauthn') {
      submitWithWebauthn();
    }
    if (mfaOption?.value === 'otp') {
      submitWithTotp(otpToken);
    }
  }

  return (
    <Validation>
      {({ validator }) => (
        <Dialog
          dialogCss={() => ({
            width: '400px',
          })}
          disableEscapeKeyDown={false}
          onClose={onClose}
          open={true}
        >
          <DialogHeader style={{ flexDirection: 'column' }}>
            <DialogTitle>Verify your identity</DialogTitle>
            <Text textAlign="center" color="text.secondary">
              You must verify your identity before peforming this action.
            </Text>
          </DialogHeader>
          {attempt.status === 'failed' && (
            <Danger mt={2} width="100%">
              {attempt.statusText}
            </Danger>
          )}
          <DialogContent>
            <Flex mt={2} alignItems="flex-end">
              <FieldSelect
                width="50%"
                label="Two-factor type"
                value={mfaOption}
                options={mfaOptions}
                onChange={(o: MfaOption) => {
                  setMfaOption(o);
                  clearAttempt();
                }}
                data-testid="mfa-select"
                mr={3}
                mb={0}
                isDisabled={attempt.status === 'processing'}
              />
              <Box width="50%">
                {mfaOption.value === 'otp' && (
                  <FieldInput
                    label="Authenticator code"
                    rule={requiredToken}
                    inputMode="numeric"
                    autoComplete="one-time-code"
                    value={otpToken}
                    onChange={e => setOtpToken(e.target.value)}
                    placeholder="123 456"
                    readonly={attempt.status === 'processing'}
                    mb={0}
                  />
                )}
              </Box>
            </Flex>
          </DialogContent>
          <DialogFooter>
            <ButtonPrimary
              onClick={e => validator.validate() && onSubmit(e)}
              disabled={attempt.status === 'processing'}
              mr={3}
              mt={3}
              autoFocus
            >
              Continue
            </ButtonPrimary>
            <ButtonSecondary onClick={onClose}>Cancel</ButtonSecondary>
          </DialogFooter>
        </Dialog>
      )}
    </Validation>
  );
}
