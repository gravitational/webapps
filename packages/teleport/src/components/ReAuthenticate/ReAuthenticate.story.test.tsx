import React from 'react';
import { render, screen } from 'design/utils/testing';
import { Loaded, Failed } from './ReAuthenticate.story';

test('render re-authentication dialog', () => {
  render(<Loaded />);

  expect(screen.getByTestId('Modal')).toMatchSnapshot();
});

test('render failed state for re-authentication dialog', () => {
  render(<Failed />);

  expect(screen.getByTestId('Modal')).toMatchSnapshot();
});
