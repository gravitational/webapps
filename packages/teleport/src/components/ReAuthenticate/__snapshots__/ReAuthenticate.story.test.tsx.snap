// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render failed state for re-authentication dialog 1`] = `
.c7 {
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 2px;
  box-sizing: border-box;
  box-shadow: 0 1px 4px rgba(0,0,0,.24);
  margin: 0 0 24px 0;
  min-height: 40px;
  padding: 8px 16px;
  overflow: auto;
  word-break: break-all;
  line-height: 1.5;
  margin-top: 8px;
  background: #f50057;
  color: #FFFFFF;
  width: 100%;
}

.c7 a {
  color: #FFFFFF;
}

.c10 {
  box-sizing: border-box;
  margin-bottom: 0px;
  margin-right: 16px;
  width: 50%;
}

.c13 {
  box-sizing: border-box;
  width: 50%;
}

.c15 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #512FC9;
  color: rgba(255,255,255,0.87);
  min-height: 32px;
  font-size: 12px;
  padding: 0px 24px;
  margin-right: 16px;
  margin-top: 16px;
}

.c15:active {
  opacity: 0.56;
}

.c15:hover,
.c15:focus {
  background: #651FFF;
}

.c15:active {
  background: #354AA4;
}

.c15:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c16 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #222C59;
  color: rgba(255,255,255,0.87);
  min-height: 32px;
  font-size: 12px;
  padding: 0px 24px;
}

.c16:active {
  opacity: 0.56;
}

.c16:hover,
.c16:focus {
  background: #2C3A73;
}

.c16:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c11 {
  color: #FFFFFF;
  display: block;
  font-size: 11px;
  font-weight: 500;
  text-transform: uppercase;
  width: 100%;
  margin-bottom: 4px;
}

.c5 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 22px;
  line-height: 32px;
  text-transform: uppercase;
  margin: 0px;
  color: rgba(255,255,255,0.87);
}

.c6 {
  overflow: hidden;
  text-overflow: ellipsis;
  margin: 0px;
  color: rgba(255,255,255,0.56);
  text-align: center;
}

.c9 {
  box-sizing: border-box;
  margin-top: 8px;
  display: flex;
  align-items: flex-end;
}

.c1 {
  z-index: -1;
  position: fixed;
  right: 0;
  bottom: 0;
  top: 0;
  left: 0;
  background-color: rgba(0,0,0,0.5);
  opacity: 1;
  touch-action: none;
}

.c0 {
  position: fixed;
  z-index: 1200;
  right: 0;
  bottom: 0;
  top: 0;
  left: 0;
}

.c2 {
  height: 100%;
  outline: none;
  color: black;
  display: flex;
  align-items: center;
  justify-content: center;
  opacity: 1;
  will-change: opacity;
  transition: opacity 225ms cubic-bezier(0.4,0,0.2,1) 0ms;
}

.c3 {
  padding: 32px;
  padding-top: 24px;
  background: #1C254D;
  color: rgba(255,255,255,0.87);
  border-radius: 8px;
  box-shadow: 0 8px 32px rgba(0,0,0,0.24);
  display: flex;
  flex-direction: column;
  position: relative;
  overflow-y: auto;
  max-height: calc(100% - 96px);
  width: 400px;
}

.c4 {
  box-sizing: border-box;
  margin-bottom: 16px;
  min-height: 32px;
  display: flex;
  align-items: center;
}

.c8 {
  box-sizing: border-box;
  margin-bottom: 32px;
  flex: 1;
  display: flex;
  flex-direction: column;
}

.c14 {
  box-sizing: border-box;
}

.c12 .react-select-container {
  box-shadow: inset 0 2px 4px rgba(0,0,0,0.24);
  box-sizing: border-box;
  border: none;
  display: block;
  font-size: 14px;
  outline: none;
  width: 100%;
  color: rgba(0,0,0,0.87);
  background-color: #ffffff;
  margin-bottom: 0px;
  border-radius: 4px;
}

.c12 .react-select__control,
.c12 .react-select__control--is-focused {
  min-height: 40px;
  height: 40px;
  background-color: transparent;
  border-color: transparent;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  box-shadow: none;
}

.c12 .react-select__control:hover,
.c12 .react-select__control--is-focused:hover {
  border-color: transparent;
  cursor: pointer;
}

.c12 .react-select__option:hover {
  cursor: pointer;
  background-color: #eceff1;
}

.c12 .react-select__option--is-focused {
  background-color: #eceff1;
}

.c12 .react-select__option--is-selected {
  background-color: #cfd8dc;
  color: inherit;
}

.c12 .react-select__option--is-selected:hover {
  background-color: #cfd8dc;
}

.c12 .react-select__menu {
  margin-top: 0px;
}

.c12 .react-select__indicator-separator {
  display: none;
}

.c12 .react-select__loading-indicator {
  display: none;
}

.c12 .react-select--is-disabled .react-select__single-value,
.c12 .react-select--is-disabled .react-select__placeholder {
  color: rgba(0,0,0,0.24);
}

.c12 .react-select--is-disabled .react-select__indicator {
  color: rgba(0,0,0,0.14);
}

<div
  class="c0"
  data-testid="Modal"
>
  <div
    aria-hidden="true"
    class="c1"
    data-testid="backdrop"
  />
  <div
    class="c2"
  >
    <div
      class="c3"
      data-testid="dialogbox"
    >
      <div
        class="c4"
        style="flex-direction: column;"
      >
        <div
          class="c5"
          color="text.primary"
        >
          Verify your identity
        </div>
        <div
          class="c6"
          color="text.secondary"
        >
          You must verify your identity before peforming this action.
        </div>
      </div>
      <div
        class="c7"
        kind="danger"
        width="100%"
      >
        an error has occurred
      </div>
      <div
        class="c8"
      >
        <div
          class="c9"
        >
          <div
            class="c10"
            data-testid="mfa-select"
            width="50%"
          >
            <label
              class="c11"
              font-size="0"
            >
              Two-factor type
            </label>
            <div
              class="c12"
            >
              <div
                class="react-select-container css-2b097c-container"
              >
                <div
                  class="react-select__control css-yk16xz-control"
                >
                  <div
                    class="react-select__value-container react-select__value-container--has-value css-g1d714-ValueContainer"
                  >
                    <div
                      class="react-select__single-value css-1uccc91-singleValue"
                    >
                      Hardware Key
                    </div>
                    <input
                      aria-autocomplete="list"
                      class="css-62g3xt-dummyInput"
                      id="react-select-3-input"
                      readonly=""
                      tabindex="0"
                      value=""
                    />
                  </div>
                  <div
                    class="react-select__indicators css-1hb7zxy-IndicatorsContainer"
                  >
                    <span
                      class="react-select__indicator-separator css-1okebmr-indicatorSeparator"
                    />
                    <div
                      aria-hidden="true"
                      class="react-select__indicator react-select__dropdown-indicator css-tlfecz-indicatorContainer"
                    >
                      <svg
                        aria-hidden="true"
                        class="css-6q0nyr-Svg"
                        focusable="false"
                        height="20"
                        viewBox="0 0 20 20"
                        width="20"
                      >
                        <path
                          d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="c13"
            width="50%"
          />
        </div>
      </div>
      <div
        class="c14"
      >
        <button
          class="c15"
          kind="primary"
        >
          Continue
        </button>
        <button
          class="c16"
          kind="secondary"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`render re-authentication dialog 1`] = `
.c9 {
  box-sizing: border-box;
  margin-bottom: 0px;
  margin-right: 16px;
  width: 50%;
}

.c12 {
  box-sizing: border-box;
  width: 50%;
}

.c14 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #512FC9;
  color: rgba(255,255,255,0.87);
  min-height: 32px;
  font-size: 12px;
  padding: 0px 24px;
  margin-right: 16px;
  margin-top: 16px;
}

.c14:active {
  opacity: 0.56;
}

.c14:hover,
.c14:focus {
  background: #651FFF;
}

.c14:active {
  background: #354AA4;
}

.c14:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c15 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #222C59;
  color: rgba(255,255,255,0.87);
  min-height: 32px;
  font-size: 12px;
  padding: 0px 24px;
}

.c15:active {
  opacity: 0.56;
}

.c15:hover,
.c15:focus {
  background: #2C3A73;
}

.c15:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c10 {
  color: #FFFFFF;
  display: block;
  font-size: 11px;
  font-weight: 500;
  text-transform: uppercase;
  width: 100%;
  margin-bottom: 4px;
}

.c5 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 22px;
  line-height: 32px;
  text-transform: uppercase;
  margin: 0px;
  color: rgba(255,255,255,0.87);
}

.c6 {
  overflow: hidden;
  text-overflow: ellipsis;
  margin: 0px;
  color: rgba(255,255,255,0.56);
  text-align: center;
}

.c8 {
  box-sizing: border-box;
  margin-top: 8px;
  display: flex;
  align-items: flex-end;
}

.c1 {
  z-index: -1;
  position: fixed;
  right: 0;
  bottom: 0;
  top: 0;
  left: 0;
  background-color: rgba(0,0,0,0.5);
  opacity: 1;
  touch-action: none;
}

.c0 {
  position: fixed;
  z-index: 1200;
  right: 0;
  bottom: 0;
  top: 0;
  left: 0;
}

.c2 {
  height: 100%;
  outline: none;
  color: black;
  display: flex;
  align-items: center;
  justify-content: center;
  opacity: 1;
  will-change: opacity;
  transition: opacity 225ms cubic-bezier(0.4,0,0.2,1) 0ms;
}

.c3 {
  padding: 32px;
  padding-top: 24px;
  background: #1C254D;
  color: rgba(255,255,255,0.87);
  border-radius: 8px;
  box-shadow: 0 8px 32px rgba(0,0,0,0.24);
  display: flex;
  flex-direction: column;
  position: relative;
  overflow-y: auto;
  max-height: calc(100% - 96px);
  width: 400px;
}

.c4 {
  box-sizing: border-box;
  margin-bottom: 16px;
  min-height: 32px;
  display: flex;
  align-items: center;
}

.c7 {
  box-sizing: border-box;
  margin-bottom: 32px;
  flex: 1;
  display: flex;
  flex-direction: column;
}

.c13 {
  box-sizing: border-box;
}

.c11 .react-select-container {
  box-shadow: inset 0 2px 4px rgba(0,0,0,0.24);
  box-sizing: border-box;
  border: none;
  display: block;
  font-size: 14px;
  outline: none;
  width: 100%;
  color: rgba(0,0,0,0.87);
  background-color: #ffffff;
  margin-bottom: 0px;
  border-radius: 4px;
}

.c11 .react-select__control,
.c11 .react-select__control--is-focused {
  min-height: 40px;
  height: 40px;
  background-color: transparent;
  border-color: transparent;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  box-shadow: none;
}

.c11 .react-select__control:hover,
.c11 .react-select__control--is-focused:hover {
  border-color: transparent;
  cursor: pointer;
}

.c11 .react-select__option:hover {
  cursor: pointer;
  background-color: #eceff1;
}

.c11 .react-select__option--is-focused {
  background-color: #eceff1;
}

.c11 .react-select__option--is-selected {
  background-color: #cfd8dc;
  color: inherit;
}

.c11 .react-select__option--is-selected:hover {
  background-color: #cfd8dc;
}

.c11 .react-select__menu {
  margin-top: 0px;
}

.c11 .react-select__indicator-separator {
  display: none;
}

.c11 .react-select__loading-indicator {
  display: none;
}

.c11 .react-select--is-disabled .react-select__single-value,
.c11 .react-select--is-disabled .react-select__placeholder {
  color: rgba(0,0,0,0.24);
}

.c11 .react-select--is-disabled .react-select__indicator {
  color: rgba(0,0,0,0.14);
}

<div
  class="c0"
  data-testid="Modal"
>
  <div
    aria-hidden="true"
    class="c1"
    data-testid="backdrop"
  />
  <div
    class="c2"
  >
    <div
      class="c3"
      data-testid="dialogbox"
    >
      <div
        class="c4"
        style="flex-direction: column;"
      >
        <div
          class="c5"
          color="text.primary"
        >
          Verify your identity
        </div>
        <div
          class="c6"
          color="text.secondary"
        >
          You must verify your identity before peforming this action.
        </div>
      </div>
      <div
        class="c7"
      >
        <div
          class="c8"
        >
          <div
            class="c9"
            data-testid="mfa-select"
            width="50%"
          >
            <label
              class="c10"
              font-size="0"
            >
              Two-factor type
            </label>
            <div
              class="c11"
            >
              <div
                class="react-select-container css-2b097c-container"
              >
                <div
                  class="react-select__control css-yk16xz-control"
                >
                  <div
                    class="react-select__value-container react-select__value-container--has-value css-g1d714-ValueContainer"
                  >
                    <div
                      class="react-select__single-value css-1uccc91-singleValue"
                    >
                      Hardware Key
                    </div>
                    <input
                      aria-autocomplete="list"
                      class="css-62g3xt-dummyInput"
                      id="react-select-2-input"
                      readonly=""
                      tabindex="0"
                      value=""
                    />
                  </div>
                  <div
                    class="react-select__indicators css-1hb7zxy-IndicatorsContainer"
                  >
                    <span
                      class="react-select__indicator-separator css-1okebmr-indicatorSeparator"
                    />
                    <div
                      aria-hidden="true"
                      class="react-select__indicator react-select__dropdown-indicator css-tlfecz-indicatorContainer"
                    >
                      <svg
                        aria-hidden="true"
                        class="css-6q0nyr-Svg"
                        focusable="false"
                        height="20"
                        viewBox="0 0 20 20"
                        width="20"
                      >
                        <path
                          d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="c12"
            width="50%"
          />
        </div>
      </div>
      <div
        class="c13"
      >
        <button
          class="c14"
          kind="primary"
        >
          Continue
        </button>
        <button
          class="c15"
          kind="secondary"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</div>
`;
